### PR TITLE
Upgraded Thanos

### PR DESCRIPTION
**Pull request by pracucci**
**What this PR does**:
In this PR I'm upgrading Thanos to bring in a couple of changes we did (see CHANGELOG). I also took the opportunity to refactor the chunks storage swift client to use the blocks storage config (this brings in a breaking change to `container-name` default value, but swift is still marked experimental).

~~**Do not merge**: I reviewed all Thanos changes and I've found a race condition introduced by https://github.com/thanos-io/thanos/pull/3557. I need to fix it merge merging this PR.~~ (Fixed)

Most Thanos changes are small and safe things. This optimisation PR https://github.com/thanos-io/thanos/pull/3531 is worth noting: diff is not easy to review but I did my best to compare before/after and changes LGTM.

**Which issue(s) this PR fixes**:
N/A

**Checklist**
- [ ] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by pracucci** [comment]
> Do not merge: I reviewed all Thanos changes and I've found a race condition introduced by thanos-io/thanos#3557. I need to fix it merge merging this PR.

This PR introduces a lock to fix it:
https://github.com/thanos-io/thanos/pull/3705

**Comment by pstibrany** [comment]
Cortex changes LGTM. Amazing that you could spot a new race bug in that huge Thanos diff!

File: CHANGELOG.md [hunk 1]
```diff
@@ -25,10 +26,16 @@
...
+  - Ruler: `-ruler.storage.swift.auth-version`, `-ruler.storage.swift.max-retries`, `-ruler.storage.swift.connect-timeout`, `-ruler.storage.swift.request-timeout`.
 * [BUGFIX] Allow `-querier.max-query-lookback` use `y|w|d` suffix like deprecated `-store.max-look-back-period`. #3598
 * [BUGFIX] Memberlist: Entry in the ring should now not appear again after using "Forget" feature (unless it's still heartbeating). #3603
 * [BUGFIX] Ingester: do not close idle TSDBs while blocks shipping is in progress. #3630
 * [BUGFIX] Ingester: correctly update `cortex_ingester_memory_users` and `cortex_ingester_active_series` when a tenant's idle TSDB is closed, when running Cortex with the blocks storage. #3646
+* [BUGFIX] Compactor: delete `deletion-mark.json` at last when deleting a block in order to not leave partial blocks without deletion mark in the bucket if the compactor is interrupted while deleting a block. #3660
+* [BUGFIX] Blocks storage: do not cleanup a partially uploaded block when `meta.json` upload fails. #3660
```
Comment (pstibrany) [inline CHANGELOG.md:38]:
```suggestion
* [BUGFIX] Blocks storage: do not cleanup a partially uploaded block when `meta.json` upload fails. Despite failure to upload `meta.json`, this file may in some cases still appear in the bucket later. By skipping early cleanup, we avoid having corrupted blocks in the storage. #3660
```

**Comment by pracucci** [comment]
@pstibrany Thanks for your review. The Thanos fix thanos-io/thanos#3705 has been merged and I've upgraded Thanos again, so should be ready for another review.

**Review by pstibrany — APPROVED** [review]

**Pull request merged**